src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -180,16 +183,52 @@ class DependencyWaiterImpl : public DependencyWaiter {
   RayletClient &raylet_client_;
 };
 
+/// Wraps a thread-pool to block posts until the pool has free slots. This is used
+/// by the SchedulingQueue to provide backpressure to clients.
+class BoundedExecutor {
+ public:
+  BoundedExecutor(int max_concurrency)
+      : num_running_(0), max_concurrency_(max_concurrency), pool_(max_concurrency){};
+
+  /// Posts work to the pool, blocking if no free threads are available.
+  void PostBlocking(std::function<void()> fn) {
+    mu_.LockWhen(absl::Condition(this, &BoundedExecutor::ThreadsAvailable));
+    num_running_ += 1;
+    mu_.Unlock();
+    boost::asio::post(pool_, [this, fn]() {
+      fn();
+      absl::MutexLock lock(&mu_);
+      num_running_ -= 1;
+    });
+  }
+
+ private:
+  bool ThreadsAvailable() EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    return num_running_ < max_concurrency_;
+  }
+
+  /// Protects access to the counters below.
+  absl::Mutex mu_;
+  /// The number of currently running tasks.
+  int num_running_ GUARDED_BY(mu_);
+  /// The max number of concurrently running tasks allowed.
+  int max_concurrency_ GUARDED_BY(mu_);
+  /// The underlying thread pool for running tasks.
+  boost::asio::thread_pool pool_;
+};
+
 /// Used to ensure serial order of task execution per actor handle.
 /// See direct_actor.proto for a description of the ordering protocol.
 class SchedulingQueue {
  public:
   SchedulingQueue(boost::asio::io_service &main_io_service, DependencyWaiter &waiter,
+                  std::shared_ptr<BoundedExecutor> pool = nullptr,
                   int64_t reorder_wait_seconds = kMaxReorderWaitSeconds)
       : wait_timer_(main_io_service),
         waiter_(waiter),
         reorder_wait_seconds_(reorder_wait_seconds),
-        main_thread_id_(boost::this_thread::get_id()) {}
+        main_thread_id_(boost::this_thread::get_id()),
+        pool_(pool) {}
```

Comment:
so the task order from a given caller is ensured by the `SchedulingQueue`?

Reply:
The ordering guarantee sounds fundamental to the actor model. Is it possible to make sure that tasks from the same actor handle is ordered? e.g. we could use `boost::strand` for posting the tasks. I think guaranteeing the order would help to make this useful to a broader range of applications, if there are some cases that order is not important and ignoring the order can bring performance benefits, then we can add another option for that, e.g. NotTaskOrder to make it more explicit. What do you think?

Reply:
This is easy to implement, but problem is that would effectively disable concurrency when submitting from one handle.

But for the use cases I know of, it is not needed. So I think the current implementation is a better default. I would suggest adding this in the future as an option.

Reply:
OK, probably we can make it more explicit that order is not guaranteed in this case, to avoid potential user confusions?